### PR TITLE
Issue 4720: Reduce cpu use on client

### LINKED ISSUE #4720: Lower cpu use in the client.
**Problem description**
Attached are some flame graphs showing some hotspots containing unnessicary work

cpu-flame-graph-100b-1s-pub

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**  
Attempts to reduce CPU use on the client.

**Purpose of the change**  
Lower cpu use at hotspots.
Fixes #4720 #4697

**What the code does**  
No functional changes. Just optimizations.

**How to verify it**  
All tests should pass, perfromance should improve.

### CODE REVIEW

**Comment by eolivelli** [comment]
How did you find this hotspots? 
With some profiling tool or just by a static analysis?

File: common/src/main/java/io/pravega/common/ExponentialMovingAverage.java [hunk 1]
```diff
@@ -20,19 +20,19 @@
 
     private final double newSampleWeight;
     private final AtomicLong valueEncodedAsLong;
-    private final boolean logarithmicWeighting;
```
Comment (RaulGracia) [inline common/src/main/java/io/pravega/common/ExponentialMovingAverage.java:23]: Is this going to reduce the CPU usage of the client?

File: client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java [hunk 2]
```diff
@@ -106,7 +106,7 @@ public String toString() {
         return ToStringUtils.mapToString(ownedSegments);
     }
 
-    private static class PositionBuilder implements ObjectBuilder<PositionImpl> {
```
Comment (RaulGracia) [inline client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java:109]: Can we add a constructor of this `PositionImpl.java` that does not execute eagerly a `for` inside the constructor? Can this `for` be executed just if the user uses this object?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java [hunk 3]
```diff
@@ -455,7 +463,7 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
     @Override
     public void write(PendingEvent event) {
         //State is set to sealed during a Transaction abort and the segment writer should not throw an {@link IllegalStateException} in such a case.
-        checkState(NameUtils.isTransactionSegment(segmentName) || !state.isAlreadySealed(), "Segment: %s is already sealed", segmentName);
```
Comment (RaulGracia) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java:458]: We have measured this to provide some small improvement in throughput (e.g., 10K e/s), so it is good to include in this PR.

**Comment by eolivelli** [comment]
Thanks for the pointer

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java [hunk 4]
```diff
@@ -248,17 +249,22 @@ private long addToInflight(PendingEvent event) {
          */
         private List<PendingEvent> removeInflightBelow(long ackLevel) {
             synchronized (lock) {
-                ConcurrentNavigableMap<Long, PendingEvent> acked = inflight.headMap(ackLevel, true);
-                List<PendingEvent> result = new ArrayList<>(acked.values());
-                acked.clear();
+                List<PendingEvent> result = new ArrayList<>();
+                Entry<Long, PendingEvent> entry = inflight.peekFirst();
+                while (entry != null && entry.getKey() <= ackLevel) {
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java:254]: Can this be rewritten as?

```
Entry<Long, PendingEvent> entry;
while(!inflight.isEmpty() && (entry = inflight.peekFirst()) != null) {
    result.add(entry.getValue);
    inflight.pollFirst();
}
```

The goal here is to reduce the number of times `peekFirst` appears in the code.

File: client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java [hunk 5]
```diff
@@ -370,10 +371,18 @@ public Type fetchEvent(EventPointer pointer) throws NoSuchEventException {
         }
     }
 
-    @Synchronized
     @VisibleForTesting
     List<EventSegmentReader> getReaders() {
-        return Collections.unmodifiableList(readers);
+        synchronized (readers) {            
+            return Collections.unmodifiableList(readers);
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java:377]: This is not really thread safe. `unmodifiableList` returns a wrapper for `readers`, but readers can still be modified by this object, thus causing issues for anyone who might be iterating over it at the same time.

File: client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java [hunk 6]
```diff
@@ -370,10 +371,18 @@ public Type fetchEvent(EventPointer pointer) throws NoSuchEventException {
...
+        }
+    }
+    
+    @VisibleForTesting
+    Map<Segment, Range> getRanges() {
+        synchronized (readers) {
+            return Collections.unmodifiableMap(ranges);
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java:384]: same here

File: client/src/test/java/io/pravega/client/stream/impl/EventStreamReaderTest.java [hunk 7]
```diff
@@ -85,7 +85,7 @@
     private final Consumer<Segment> segmentSealedCallback = segment -> { };
     private final EventWriterConfig writerConfig = EventWriterConfig.builder().build();
 
-    @Test(timeout = 10000)
+    @Test//(timeout = 10000)
```
Comment (andreipaduroiu) [inline client/src/test/java/io/pravega/client/stream/impl/EventStreamReaderTest.java:88]: Revert this?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java (hunk 4 above)
Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java:254]: This ignores the loop conditional: `entry.getKey() <= ackLevel`

**Pull request merged**